src/haptics.cpp

```diff
@@ -12,51 +15,86 @@ init_haptics() {
   }
 }
 
+/* Vibrates specified motors for given frequency.
+ *
+ * @param motor_idx	    specified motor to vibrate
+ * @param duration	    length time of vibration in ms
+ * @param percent_motor percent of haptic vibration capacity to vibrate at
+ * @param frequency     frequency of vibration (Hz)
+ */
+void actuate_hz(int motor_idx, int duration, double percent_motor, float frequency) {
+  percent_motor = percent_motor * 0.85;
+  if (percent_motor > 0.85) {
+    Serial.println("Motor percentage above 85%. Using 85%");
+    percent_motor = 0.85;
+  }
+
+  if (frequency == 0) {
+    delayMicroseconds(duration * 1000);
+    return;
+  }
+  
+  float half_period = 1 / (frequency * 2);
+  int num_cycles = duration / (2 * half_period * 1000);
+  digitalWrite(MOTOR_PINS[motor_idx] + EN_OFFSET, HIGH);  // enable
+
+  for (int j = 0; j < num_cycles; j++) {
+    analogWrite(MOTOR_PINS[motor_idx], 512 + percent_motor * 512);
+    delayMicroseconds(half_period * MICROSECONDS_PER_SECOND);
+    analogWrite(MOTOR_PINS[motor_idx], 512 - percent_motor * 512);
+    delayMicroseconds(half_period * 1000000);
+  }
+  analogWrite(MOTOR_PINS[motor_idx], 512 + percent_motor * 512);
+  delayMicroseconds(half_period / 2 * 1000000);
+  analogWrite(MOTOR_PINS[motor_idx], 512);
+  digitalWrite(MOTOR_PINS[motor_idx] + EN_OFFSET, LOW);  // disable
+}
+
 /* Vibrates specified motors for given intensity and duration.
  *
- * @param motor_pin	specified motor to vibrate
+ * @param motor_idx	specified motor to vibrate
  * @param duration	length time of vibration in ms
  * @param percent_motor percent of haptic vibration capacity to vibrate at,
- * 			not to exceed 0.85
  */
-void actuate_motor(int motor_pin, int duration, double percent_motor) {
-  digitalWrite(MOTOR_PINS[motor_pin] + EN_OFFSET, HIGH);  // enable
-  int frequency = 3;
+void actuate_motor(int motor_idx, int duration, double percent_motor) {
+  digitalWrite(MOTOR_PINS[motor_idx] + EN_OFFSET, HIGH);  // enable
+  percent_motor = percent_motor * 0.85;
   if (percent_motor > 0.85) {
```

Comment:
Because we multiply the input by 0.85, I think better input validation/ error handling would be to check if `percent_motor` (the input) is `0 < x < 1`.

Also still don't understand why `percent_motor` is set back to 1 on line 66. Is this a mistake?